File: Ordination/Python/BubbleSort.py
#coding: utf-8
from Sort import Sort

class BubbleSort(Sort):
	pass

	'''
		Generic Sort Method
		algorithm.sort(array) -> return  complete ordered array
		algorithm.sort(array,leftLimit,rightLimit) -> return partially ordered array
	'''
	def sort(self,lista,leftIndex = 0,rightIndex = None):
		if rightIndex == None:
			rightIndex = len(lista) - 1
		
		if self.__validateParams(lista,leftIndex,rightIndex):
			#self.__simpleBubbleSort(lista,leftIndex,rightIndex)
			#self.__bidirectionalBubbleSort(lista,leftIndex,rightIndex)
			self.__recursiveBidirectionalBubbleSort(lista,leftIndex,rightIndex,True)
	
	'''
		Simple Iterative BubbleSort
	'''
	def __simpleBubbleSort(self,lista,leftIndex,rightIndex):
		flag = True
		while(flag):
			flag = False;
			for i in range(leftIndex,rightIndex):
				if(lista[i] > lista[i+1]):
					lista[i] , lista[i +1] = lista[i+1] , lista[i]
					flag = True
	'''
		Iterative Bidirectional BubbleSort in place
	'''
	def __bidirectionalBubbleSort(self,lista,leftIndex,rightIndex):
		flag = True
		while(flag  and leftIndex < rightIndex):
			flag = False;
			for i in range(leftIndex,rightIndex):
				if(lista[i] > lista[i+1]):
					lista[i],lista[i+1] = lista[i+1] , lista[i]
					flag = True
			flag = False
			rightIndex = rightIndex- 1
			for i in range(rightIndex,leftIndex +1,-1):
				if(lista[i] < lista[i - 1]):
					lista[i-1],lista[i] = lista[i] , lista[i-1]
					flag = True
			leftIndex = leftIndex +1
	'''
		Recursive Bididrectional BubbleSort in place
	'''
	def __recursiveBidirectionalBubbleSort(self,lista,leftIndex,rightIndex,flag):
		if(flag and leftIndex < rightIndex):
			flag = False
			for i in range(leftIndex,rightIndex):
				if(lista[i] > lista[i+1]):
					lista[i] ,lista[i + 1] = lista[i+1] ,lista[i]
					flag = True
			flag = False
			rightIndex = rightIndex -1
			for i in range(rightIndex,leftIndex,-1):
				if(lista[i] < lista[i-1]):
					lista[i] , lista[i - 1] = lista[i-1],lista[i]
					flag = True
			leftindex = leftIndex +1
			self.__recursiveBidirectionalBubbleSort(lista,leftIndex,rightIndex,flag)
	
	''''
		This method verify if parameters are valid if some parameter is not valid the method returns false
	'''
	def __validateParams(self,lista , leftIndex , rightIndex):
		if(lista == None  or leftIndex < 0 or rightIndex > len(lista) - 1 or leftIndex >= rightIndex):
			return False
		else:
			return True

File: Ordination/C++/BubbleSort.cpp
#include <iostream>

template <typename T> class BubbleSort{
	public:
		void sort(T array[],int leftIndex,int rightIndex,int size){
			if(validateParams(array,leftIndex,rightIndex,size)){
				//simpleBubbleSort(array,leftIndex,rightIndex);
				//bidirectionalBubbleSort(array,leftIndex,rightIndex);
				this->recursiveBidirectionalBubbleSort(array,leftIndex,rightIndex,true);
				this->showArray(array,size);
			}else{
				std::cout << "Nao esta entrando!!" << std::endl;
				
			}	
		}
		void sort(T array[],int size){
			this->sort(array,0,size-1,size);
		}
	private:
		bool validateParams(T array[],int leftIndex,int rightIndex,int size){
			if(array == NULL || leftIndex < 0 || rightIndex > size - 1 || leftIndex >= rightIndex){
				return false;
			}else{	
				return true;
			}
		}
	
		void simpleBubbleSort(T array[],int leftIndex,int rightIndex){
			int ordered  = 0;
			for(int i = leftIndex ; i < rightIndex;i++){
				bool flag = false;
				for(int j = leftIndex; j < rightIndex-ordered;j++){
					if(array[j] > array[j+1]){
						swap(array,j,j+1);
						flag = true;
					}
				}
				ordered++;
				if(!flag){
					break;
				}
			}
		}

		void bidirectionalBubbleSort(T array[],int leftIndex,int rightIndex){
			bool flag= true;
			while(flag){
				flag = false;
				for(int i = leftIndex;i < rightIndex;i++){
					if(array[i] > array[i+1]){
						swap(array,i,i+1);
					}
				}
				rightIndex--;
				flag = false;
				for(int i = rightIndex; i > leftIndex;i--){
					if(array[i] < array[i-1]){
						swap(array,i,i-1);
						flag = true;
					}
				}
				leftIndex++;
			}
		}
		
		void recursiveBidirectionalBubbleSort(T array[],int leftIndex,int rightIndex,bool flag){
			if(flag){
				flag = false;
				for(int i = 0 ; i < rightIndex; i++){
					if(array[i] > array[i+1]){
						swap(array,i,i+1);
						flag = true;
					}
				}
				rightIndex--;
				
				flag = false;
				for(int i = rightIndex; i > leftIndex;i--){
					if(array[i] < array[i-1]){
						swap(array,i,i-1);
						flag = true;
					}
				}
				leftIndex++;
				this->recursiveBidirectionalBubbleSort(array,leftIndex,rightIndex,flag);
			}	
		}
		void swap(T array[],int i,int j){
			T aux = array[i];
			array[i] = array[j];
			array[j] = aux;
		}
		
		void showArray(T array[],int size){
			std::cout << "Array: [";
			for(int i = 0 ; i < size;i++){
				if(i != size- 1){
					std::cout << array[i] << ",";
				}else{
					std::cout << array[i] << "]";
				}
			}	
			std::cout << std::endl;
		}
};
